advanced_alchemy/service/_async.py

```diff
@@ -0,0 +1,526 @@
+    from collections.abc import Sequence
+
+    from sqlalchemy import Select, StatementLambdaElement
+    from sqlalchemy.ext.asyncio import AsyncSession
+    from sqlalchemy.orm import InstrumentedAttribute
+    from sqlalchemy.sql import ColumnElement
+
+    from advanced_alchemy.filters import FilterTypes
+    from advanced_alchemy.repository import SQLAlchemyAsyncRepository
+    from advanced_alchemy.service.typing import FilterTypeT
+
+
+class SQLAlchemyAsyncRepositoryService(Generic[ModelT]):
+    """Service object that operates on a repository object."""
+
+    repository_type: type[SQLAlchemyAsyncRepository[ModelT]]
+    match_fields: list[str] | None = None
+
+    def __init__(
+        self,
+        session: AsyncSession,
+        statement: Select[tuple[ModelT]] | StatementLambdaElement | None = None,
+        auto_expunge: bool = False,
+        auto_refresh: bool = True,
+        auto_commit: bool = False,
+        **repo_kwargs: Any,
+    ) -> None:
+        """Configure the service object.
+
+        Args:
+            session: Session managing the unit-of-work for the operation.
+            statement: To facilitate customization of the underlying select query.
+            auto_expunge: Remove object from session before returning.
+            auto_refresh: Refresh object from session before returning.
+            auto_commit: Commit objects before returning.
+            **repo_kwargs: passed as keyword args to repo instantiation.
+        """
+        self.repository = self.repository_type(
+            session=session,
+            statement=statement,
+            auto_expunge=auto_expunge,
+            auto_refresh=auto_refresh,
+            auto_commit=auto_commit,
+            **repo_kwargs,
+        )
+
+    async def count(
+        self,
+        *filters: FilterTypes,
+        statement: Select[tuple[ModelT]] | StatementLambdaElement | None = None,
+        **kwargs: Any,
+    ) -> int:
+        """Count of records returned by query.
+
+        Args:
+            *filters: arguments for filtering.
+            statement: To facilitate customization of the underlying select query.
+                Defaults to :class:`SQLAlchemyAsyncRepository.statement <SQLAlchemyAsyncRepository>`
+            **kwargs: key value pairs of filter types.
+
+        Returns:
+           A count of the collection, filtered, but ignoring pagination.
+        """
+        return await self.repository.count(*filters, statement=statement, **kwargs)
+
+    async def create(self, data: ModelT | dict[str, Any]) -> ModelT:
+        """Wrap repository instance creation.
+
+        Args:
+            data: Representation to be created.
+
+        Returns:
+            Representation of created instance.
+        """
+        data = await self.to_model(data, "create")
+        return await self.repository.add(data)
+
+    async def create_many(
+        self,
+        data: list[ModelT | dict[str, Any]] | list[dict[str, Any]] | list[ModelT],
```

Comment:
Can this and the other occurrences be replaced with `ModelDictListT` alias defined in `advanced_alchemy/service/typing.py`?

```suggestion
        data: ModelDictListT
```

Reply:
It ended up complaining in a few other places on the repository then.  I'm sure there's a way to clean these up though.